From 203548fa6e72381618eb7d1711ab07b86a62e4a7 Mon Sep 17 00:00:00 2001
From: Neal Frager <neal.frager@amd.com>
Date: Fri, 24 Feb 2023 13:53:09 +0000
Subject: [PATCH v1 1/1] pmufw: pm_sram: use 32-bit writes for tcm ecc init

The memset function defaults to 8-bit transactions on some compilers.  Since
the TCM memory ECC requires data to be written 32 bits at a time during
initialization, using memset could cause ecc errors depending on the compiler
configuration.

This patch makes sure that the tcm ecc initialization always succeeds
regardless of the compiler used.

Signed-off-by: Neal Frager <neal.frager@amd.com>
---
 lib/sw_apps/zynqmp_pmufw/src/pm_sram.c | 17 +++++++++++++++--
 1 file changed, 15 insertions(+), 2 deletions(-)

diff --git a/lib/sw_apps/zynqmp_pmufw/src/pm_sram.c b/lib/sw_apps/zynqmp_pmufw/src/pm_sram.c
index f3f9d6ede6..cac972c36f 100644
--- a/lib/sw_apps/zynqmp_pmufw/src/pm_sram.c
+++ b/lib/sw_apps/zynqmp_pmufw/src/pm_sram.c
@@ -169,7 +169,14 @@ done:
  */
 static void PmTcm0EccInit(const PmSlaveTcm* const tcm)
 {
-	(void)memset((u32 *)tcm->base, (s32)0U, tcm->size);
+	u32 base = tcm->base;
+	u32 size = tcm->size;
+
+	while(size > 0) {
+		Xil_Out32(base, 0);
+		base += 4;
+		size -= 4;
+	}
 }
 
 /**
@@ -180,11 +187,17 @@ static void PmTcm1EccInit(const PmSlaveTcm* const tcm)
 {
 	u32 base = tcm->base;
 	u32 ctrl = XPfw_Read32(RPU_RPU_GLBL_CNTL);
+	u32 size = tcm->size;
 
 	if (0U != (ctrl & RPU_RPU_GLBL_CNTL_TCM_COMB_MASK)) {
 		base -= 0x80000U;
 	}
-	(void)memset((u32 *)base, (s32)0U, tcm->size);
+
+	while(size > 0) {
+		Xil_Out32(base, 0);
+		base += 4;
+		size -= 4;
+	}
 }
 
 /**
-- 
2.17.1

